File: tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        diameter = [0]
        self.height(root, diameter)
        return diameter[0]
    
    def height(self, root:Optional[TreeNode], diameter: List) -> int:
        if not root:
            return 0
        
        lh = self.height(root.left, diameter)
        rh = self.height(root.right, diameter)
        diameter[0] = max(diameter[0], lh+rh)
        return max(lh, rh)+1
        

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.go
func max(x int, y int)int{
    if x>y {
        return x
    }
    return y
}
func calc(root *TreeNode, diameter *int) int{
    if root == nil {
        return 0
    }
    lh := calc(root.Left, diameter)
    rh := calc(root.Right, diameter)
    *diameter = max(*diameter, lh+rh)
    return 1+max(lh, rh)
}
func diameterOfBinaryTree(root *TreeNode) int {
    diameter :=0
    calc(root, &diameter)
    return diameter
}